### DISCUSSION TITLE
Setting Up A Basic Structure For Function Scheduling

### DISCUSSION THREAD

**Discussion by tyamut** [django-q2/django-q2#158] [Q&A]
I'm kind of new to using Django different then its basic scope. I scheduled a function that works every minute for trial purposes. 

In my attempt, I created schedule.py as below script and to make it auto-work when I use manage.py qcluster I defined 'ready' method in apps.py. It seems to work but even though I read the documentation many times it is still obscure for me that what is the right way of doing the scheduling.

Furthermore, in my case the earlier tasks that I run are still showing up everytime I run qcluster despite I deleted them from the Tasks object.

Can you please provide a clarification for me. Note that, I'm using DjangoORM.

```python
def schedule_price_list_status():
    print('PRICE LIST SCHEDULED')

def setup_periodic_tasks(**kwargs):
    print('SETUP TASKS')
    schedule('my_app.schedules.schedule_price_list_status', schedule_type='I', repeats=2, minutes=1)

setup_periodic_tasks()

**Comment by GDay** [comment]
You only have to define a schedule once. The best way to do that is by one of these options:

- Create a migration script that will create the schedule in the database. Migration scripts are a one-off thing when deploying.
- Create it dynamically through the code. Sometimes you might want to set up a schedule for a specific person or for an organization or so. In those cases you could addd the `schedule` command in any of the views where you want it be set up when the user clicks on a button or so.
- Manually create them through the admin portal. Go to Django_q2 -> scheduled tasks and click on "add scheduled task".

> I use manage.py qcluster I defined 'ready' method in apps.py

I wouldn't recommend that, this could create duplicate items of the schedules.

> Furthermore, in my case the earlier tasks that I run are still showing up everytime I run qcluster despite I deleted them from the Tasks object.

You would need to go to Django Q2 -> scheduled tasks and delete the task there. There might be multiple duplicated ones if you have been using the script from above.

**Reply by kennyhei** [reply]
@GDay The listed options work but for more global scheduled tasks (where migration or admin portal is the preferred way) it might be confusing for devs that where do these scheduled tasks originate from.

Just a suggestion but for this purpose it would be neat if in the future we could explicitly write scheduled task setups. What I mean is that we could have e.g. `@schedule` decorator which we can use to setup a scheduled task:

```python
@schedule(schedule_type=Schedule.HOURLY)
def reports_hourly():
   send_report()
```

This way it would be easy to track from codebase that what kind of scheduled tasks have been configured for the app.

**EDIT:** Very simple and quick sample:

```python
from datetime import datetime
from django_q.conf import croniter
from django_q.models import Schedule
from django_q.tasks import schedule
from django_q.utils import get_func_repr, localtime

def schedule_cron(cron, name=None):
    def decorator(func):
        func_name = get_func_repr(func)
        schedule_name = name or func_name
        if Schedule.objects.filter(name=schedule_name).exists():
            return
        schedule(
            func_name,
            schedule_type=Schedule.CRON,
            cron=cron,
            name=schedule_name,
            # don't run immediately, instead calculate next run based on cron
            next_run=croniter(cron, localtime()).get_next(datetime),
        )
    return decorator

# remember to install croniter first:
# https://django-q2.readthedocs.io/en/master/install.html#croniter-package
# run every hour at minute 0 -> e.g. 09:00, 10:00 etc.
@schedule_cron("0 * * * *")
def hello_world():
    return "HELLO WORLD!"
```

Tested it and it works! Although you need to be a little careful that you don't put the decorator there before you have finished writing code in case the scheduler starts to run it 😅 But shouldn't be too big of a problem in local environment.

Also not yet sure how django-q2 handles situation where `hello_world` is deleted afterwards and the schedule object stays in the scheduler forever due to `repeats=-1`. Then again, with this decorator you could create a scheduled task to cleanup all scheduled objects that have had `X` failed tasks 🤔

**Reply by GDay** [reply]
Yeah, that would be nice. At some point, I will likely integrate this package that already does all of this: https://github.com/westerveltco/django-q-registry. You could use it if you need that usecase now.

**Reply by kennyhei** [reply]
Oh okay, nice! Haven't heard about that package before but I will certainly try it, thanks 👏
